sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -236,24 +236,38 @@ case class In(value: Expression, list: Seq[Expression]) extends Predicate {
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val valueGen = value.genCode(ctx)
     val listGen = list.map(_.genCode(ctx))
+    ctx.addMutableState("boolean", ev.value, "")
+    ctx.addMutableState("boolean", ev.isNull, "")
+    val valueArg = ctx.freshName("valueArg")
```

Comment:
why do you need this? Can't you use simply `valueGen.value`?

Reply:
If `valueGen.value` is not global variable, `valueGen.value` is out of the scope in the generated method for splitting.

Reply:
if you pass it as an argument to the function you have it in scope though, exactly as you are doing with the newly generated variable.

Reply:
`valueGen.value` may have casted variable (i.e. (UTF8String)variable1). It cannot be passed as an argument.

Reply:
sorry but I can't understand your last comment. Might you please explain me?

Reply:
At [this line](https://github.com/apache/spark/pull/19733/files#diff-aec2bffff270efab6835320c7a1dc8e9R255), we cannot use `valueGen.value` in some cases (e.g. ` (UTF8String)variable1`). Then, we cannot use `valueGen.value` at [this line](https://github.com/apache/spark/pull/19733/files#diff-aec2bffff270efab6835320c7a1dc8e9R248).

Reply:
Oh, sorry, I wasn't aware that `valueGen.value` can contains things like that. Thanks for the explanation.